alf/examples/train_ppo_pr2.py

```diff
@@ -0,0 +1,327 @@
+                     'Number of environment steps to run before finishing.')
+flags.DEFINE_integer('num_epochs', 25,
+                     'Number of epochs for computing policy updates.')
+flags.DEFINE_integer(
+    'collect_episodes_per_iteration', 30,
+    'The number of episodes to take in the environment before '
+    'each update. This is the total across all parallel '
+    'environments.')
+flags.DEFINE_integer('num_eval_episodes', 30,
+                     'The number of episodes to run eval on.')
+flags.DEFINE_boolean('use_rnns', False,
+                     'If true, use RNN for policy and value function.')
+flags.DEFINE_multi_string('gin_file', None, 'Paths to the gin-config files.')
+flags.DEFINE_multi_string('gin_param', None, 'Gin binding parameters.')
+FLAGS = flags.FLAGS
+
+
+@gin.configurable
+def train_eval(
+    root_dir,
+    env_name='SocialBot-Pr2Gripper-v0',
+    env_load_fn=suite_socialbot.load,
+    random_seed=0,
+    # TODO(b/127576522): rename to policy_fc_layers.
+    actor_fc_layers=(100, 50, 25),
+    value_fc_layers=(100, 50, 25),
+    conv_layer_params=None, #[(32,8,4),(64,4,2),(64,3,1)],
+    use_rnns=False,
+    # Params for collect
+    num_environment_steps=10000000,
+    collect_episodes_per_iteration=30,
+    num_parallel_environments=30,
+    replay_buffer_capacity=1001,  # Per-environment
+    # Params for train
+    num_epochs=25,
+    learning_rate=1e-4,
+    entropy_regularization=0.002,
+    # Params for eval
+    num_eval_episodes=30,
+    eval_interval=1000,
+
+    train_checkpoint_interval=1000,
+    policy_checkpoint_interval=1000,
+
+    # Params for summaries and absl_logging
+    log_interval=50,
+    summary_interval=50,
+    summaries_flush_secs=1,
+    use_tf_functions=True,
+    debug_summaries=True,
+    summarize_grads_and_vars=True):
+  """A simple train and eval for PPO."""
+  if root_dir is None:
+    raise AttributeError('train_eval requires a root_dir.')
+
+  root_dir = os.path.expanduser(root_dir)
+  train_dir = os.path.join(root_dir, 'train')
+  eval_dir = os.path.join(root_dir, 'eval')
+
+  train_summary_writer = tf.compat.v2.summary.create_file_writer(
+      train_dir, flush_millis=summaries_flush_secs * 1000)
+  train_summary_writer.set_as_default()
+
+  eval_summary_writer = tf.compat.v2.summary.create_file_writer(
+      eval_dir, flush_millis=summaries_flush_secs * 1000)
+  eval_metrics = [
+      tf_metrics.AverageReturnMetric(buffer_size=num_eval_episodes),
+      tf_metrics.AverageEpisodeLengthMetric(buffer_size=num_eval_episodes)
+  ]
+
+  global_step = tf.compat.v1.train.get_or_create_global_step()
+  with tf.compat.v2.summary.record_if(
+      lambda: tf.math.equal(global_step % summary_interval, 0)):
+    tf.compat.v1.set_random_seed(random_seed)
+
+    if num_parallel_environments > 1:
+      tf_env = tf_py_environment.TFPyEnvironment(
+        parallel_py_environment.ParallelPyEnvironment(
+          [lambda: env_load_fn(env_name)] * num_parallel_environments))
```

Comment:
Muti-thread env does not need to be wrapped with process, or you may found the number of running process doubled.
[lambda: suite_socialbot.load(env_name,wrap_with_process=False)] * num_parallel_environments))

Reply:
I am not sure here.. it seems python multithreading still suffers from "global lock" (IMPALA use multiprocess), also gazebo env has to be in single process, as all the sensors are global singletons, multithreading will mess up with states.